File: android/app/src/main/kotlin/com/example/amble/AmbleFlutterHost.kt
package com.example.amble

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.plugin.common.MethodChannel

/** One app-process engine and Hive cache, shared by assistant calls and UI. */
object AmbleFlutterHost {
    private val handler = Handler(Looper.getMainLooper())
    private var engine: FlutterEngine? = null
    private var channel: MethodChannel? = null
    private var ready = false
    var foreground = false

    fun engine(context: Context): FlutterEngine {
        check(Looper.myLooper() == Looper.getMainLooper())
        engine?.let { return it }
        val created = FlutterEngine(context.applicationContext)
        engine = created
        channel = MethodChannel(created.dartExecutor.binaryMessenger, "com.amble/app_intents")
        channel!!.setMethodCallHandler { call, result ->
            when (call.method) {
                "ready" -> { ready = true; result.success(null) }
                "isForeground" -> result.success(foreground)
                else -> result.notImplemented()
            }
        }
        created.dartExecutor.executeDartEntrypoint(DartExecutor.DartEntrypoint.createDefault())
        return created
    }

    /** Timeouts never retry a write whose reply may have been lost. */
    fun request(context: Context, method: String, arguments: Map<String, Any>,
                callback: (Result<Any>) -> Unit) {
        handler.post {
            try {
                engine(context)
            } catch (error: Exception) {
                callback(Result.failure(error))
                return@post
            }
            val deadline = SystemClock.elapsedRealtime() + 15_000
            val waitForReady = object : Runnable {
                override fun run() {
                    if (ready) {
                        dispatch(method, arguments, callback)
                    } else if (SystemClock.elapsedRealtime() >= deadline) {
                        callback(Result.failure(IllegalStateException(
                            "Amble is not ready. Open the app and try again.")))
                    } else {
                        handler.postDelayed(this, 50)
                    }
                }
            }
            waitForReady.run()
        }
    }

    private fun dispatch(method: String, arguments: Map<String, Any>,
                         callback: (Result<Any>) -> Unit) {
        var finished = false
        fun finish(result: Result<Any>) {
            if (finished) return
            finished = true
            callback(result)
        }
        val timeout = Runnable {
            finish(Result.failure(IllegalStateException(
                "Amble could not confirm completion. Check the app before trying again.")))
        }
        handler.postDelayed(timeout, 25_000)
        channel!!.invokeMethod(method, arguments, object : MethodChannel.Result {
            override fun success(result: Any?) {
                handler.removeCallbacks(timeout)
                if (result == null) notImplemented() else finish(Result.success(result))
            }
            override fun error(code: String, message: String?, details: Any?) {
                handler.removeCallbacks(timeout)
                finish(Result.failure(IllegalArgumentException(message ?: "Amble action failed.")))
            }
            override fun notImplemented() {
                handler.removeCallbacks(timeout)
                finish(Result.failure(IllegalStateException("This Amble action is unavailable.")))
            }
        })
    }
}

File: ios/Runner/AmbleFlutterHost.swift
import Flutter
import UIKit
import flutter_local_notifications

/// One engine for both scene-less App Intents and the visible app. A second
/// Hive engine would have its own stale in-memory cache and competing writer.
@MainActor
final class AmbleFlutterHost {
  static let shared = AmbleFlutterHost()
  private(set) lazy var engine = FlutterEngine(
    name: "amble.shared", project: nil, allowHeadlessExecution: true)
  private var started = false
  private var ready = false
  private var channel: FlutterMethodChannel?

  @discardableResult
  func start() -> Bool {
    if started { return true }
    guard engine.run() else { return false }
    started = true
    FlutterLocalNotificationsPlugin.setPluginRegistrantCallback { registry in
      GeneratedPluginRegistrant.register(with: registry)
    }
    GeneratedPluginRegistrant.register(with: engine)
    let channel = FlutterMethodChannel(
      name: "com.amble/app_intents", binaryMessenger: engine.binaryMessenger)
    self.channel = channel
    channel.setMethodCallHandler { [weak self] call, result in
      switch call.method {
      case "ready":
        self?.ready = true
        result(nil)
      case "isForeground":
        result(UIApplication.shared.applicationState == .active)
      default:
        result(FlutterMethodNotImplemented)
      }
    }
    return true
  }

  @available(iOS 16.0, *)
  func request(_ method: String, _ arguments: [String: Any] = [:]) async throws -> Any {
    guard start() else { throw AmbleIntentFailure("Amble could not start. Please open the app.") }
    let deadline = Date().addingTimeInterval(15)
    while !ready {
      try Task.checkCancellation()
      guard Date() < deadline else {
        // Nothing dispatched: a late ready signal must not perform this request.
        throw AmbleIntentFailure("Amble is not ready. Please open the app and try again.")
      }
      try await Task.sleep(nanoseconds: 50_000_000)
    }
    try Task.checkCancellation()
    guard let channel else { throw AmbleIntentFailure("Amble is not available.") }
    return try await withCheckedThrowingContinuation { continuation in
      var finished = false
      let timeout = DispatchWorkItem {
        guard !finished else { return }
        finished = true
        // Never retry a write after a lost reply: it may already have committed.
        continuation.resume(throwing: AmbleIntentFailure(
          "Amble could not confirm completion. Check the app before trying again."))
      }
      DispatchQueue.main.asyncAfter(deadline: .now() + 25, execute: timeout)
      channel.invokeMethod(method, arguments: arguments) { value in
        guard !finished else { return }
        finished = true
        timeout.cancel()
        if let error = value as? FlutterError {
          continuation.resume(throwing: AmbleIntentFailure(
            error.message ?? "Amble could not complete this action."))
        } else if let value, !(value is NSObject && (value as AnyObject) === FlutterMethodNotImplemented) {
          continuation.resume(returning: value)
        } else {
          continuation.resume(throwing: AmbleIntentFailure("This Amble action is not available."))
        }
      }
    }
  }
}
